File: test/services/backend_activation_service_test.dart
import 'dart:convert';

import 'package:crypto/crypto.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:fstapp/services/backend_activation_service.dart';

void main() {
  const tenant = 'festival-test';
  const legacyUrl = 'https://legacy.supabase.co';
  const canonicalUrl = 'https://api.festapp.net';
  final canonicalBytes = utf8.encode(
    '{"schemaVersion":1,"tenantId":"$tenant","generation":1,'
    '"backend":"canonical"}\n',
  );
  final canonicalSha = sha256.convert(canonicalBytes).toString();

  BackendActivationService service({
    required Map<String, String> storage,
    required Future<List<int>> Function(Uri) fetch,
  }) =>
      BackendActivationService(
        tenantId: tenant,
        manifestUrl: 'https://app.example.test/backend-activation.json',
        canonicalManifestSha256: canonicalSha,
        legacySupabaseUrl: legacyUrl,
        legacyAnonKey: 'legacy-key',
        canonicalSupabaseUrl: canonicalUrl,
        canonicalAnonKey: 'canonical-key',
        read: (key) async => storage[key],
        write: (key, value) async => storage[key] = value,
        fetch: fetch,
      );

  test('defaults to the legacy writer when activation is unavailable',
      () async {
    final resolved = await service(
      storage: {},
      fetch: (_) async => throw Exception('offline'),
    ).resolve();

    expect(resolved.supabaseUrl, legacyUrl);
    expect(resolved.isCanonical, isFalse);
  });

  test('accepts only the exact canonical document and persists it', () async {
    final storage = <String, String>{};
    final first = await service(
      storage: storage,
      fetch: (_) async => canonicalBytes,
    ).resolve();
    expect(first.supabaseUrl, canonicalUrl);
    expect(first.isCanonical, isTrue);

    final sticky = await service(
      storage: storage,
      fetch: (_) async => throw Exception('manifest regressed'),
    ).resolve();
    expect(sticky.supabaseUrl, canonicalUrl);
    expect(sticky.isCanonical, isTrue);
  });

  test('rejects a modified or legacy activation response', () async {
    final storage = <String, String>{};
    final modified = utf8.encode(
      '{"schemaVersion":1,"tenantId":"$tenant","generation":0,'
      '"backend":"legacy"}\n',
    );
    final resolved = await service(
      storage: storage,
      fetch: (_) async => modified,
    ).resolve();

    expect(resolved.supabaseUrl, legacyUrl);
    expect(storage, isEmpty);
  });

  test('storage read and write failures keep the legacy writer', () async {
    final readFailure = BackendActivationService(
      tenantId: tenant,
      manifestUrl: 'https://app.example.test/backend-activation.json',
      canonicalManifestSha256: canonicalSha,
      legacySupabaseUrl: legacyUrl,
      legacyAnonKey: 'legacy-key',
      canonicalSupabaseUrl: canonicalUrl,
      canonicalAnonKey: 'canonical-key',
      read: (_) async => throw Exception('storage unavailable'),
      write: (_, __) async {},
      fetch: (_) async => canonicalBytes,
    );
    expect((await readFailure.resolve()).isCanonical, isFalse);

    final writeFailure = BackendActivationService(
      tenantId: tenant,
      manifestUrl: 'https://app.example.test/backend-activation.json',
      canonicalManifestSha256: canonicalSha,
      legacySupabaseUrl: legacyUrl,
      legacyAnonKey: 'legacy-key',
      canonicalSupabaseUrl: canonicalUrl,
      canonicalAnonKey: 'canonical-key',
      read: (_) async => null,
      write: (_, __) async => throw Exception('storage unavailable'),
      fetch: (_) async => canonicalBytes,
    );
    expect((await writeFailure.resolve()).isCanonical, isFalse);
  });
}

File: web_client/tests/core/backend_activation_service.test.js
import assert from 'node:assert/strict';
import { webcrypto } from 'node:crypto';
import test from 'node:test';

import { BackendActivationService } from '../../src/services/backend_activation_service.js';

const tenantId = 'festival-test';
const canonicalDocument = new TextEncoder().encode(
  `{"schemaVersion":1,"tenantId":"${tenantId}","generation":1,"backend":"canonical"}\n`,
);
const digest = Buffer.from(
  await webcrypto.subtle.digest('SHA-256', canonicalDocument),
).toString('hex');
const config = {
  supabaseUrl: 'https://legacy.supabase.co',
  anonKey: 'legacy-key',
  backendActivationTenantId: tenantId,
  backendActivationManifestUrl: 'https://app.example.test/backend-activation.json',
  backendActivationCanonicalManifestSha256: digest,
  backendActivationCanonicalSupabaseUrl: 'https://api.festapp.net',
  backendActivationCanonicalAnonKey: 'canonical-key',
};

function memoryStorage() {
  const values = new Map();
  return {
    values,
    getItem: (key) => values.get(key) ?? null,
    setItem: (key, value) => values.set(key, value),
  };
}

test('activation defaults to legacy and accepts only the pinned canonical bytes', async () => {
  const storage = memoryStorage();
  const unavailable = new BackendActivationService({
    config, storage, subtle: webcrypto.subtle,
    fetchImpl: async () => { throw new Error('offline'); },
  });
  assert.equal((await unavailable.resolve()).supabaseUrl, config.supabaseUrl);

  const activated = new BackendActivationService({
    config, storage, subtle: webcrypto.subtle,
    fetchImpl: async () => new Response(canonicalDocument),
  });
  assert.equal(
    (await activated.resolve()).supabaseUrl,
    config.backendActivationCanonicalSupabaseUrl,
  );
  assert.equal(storage.values.size, 1);
});

test('canonical activation is monotonic when the remote document disappears', async () => {
  const storage = memoryStorage();
  const marker = BackendActivationService.markerPrefix + tenantId;
  storage.setItem(marker, BackendActivationService.canonicalMarker);
  const service = new BackendActivationService({
    config, storage, subtle: webcrypto.subtle,
    fetchImpl: async () => { throw new Error('regressed'); },
  });
  assert.equal((await service.resolve()).isCanonical, true);
});

test('a modified activation response cannot select the canonical writer', async () => {
  const storage = memoryStorage();
  const service = new BackendActivationService({
    config, storage, subtle: webcrypto.subtle,
    fetchImpl: async () => new Response(canonicalDocument.slice(0, -1)),
  });
  assert.equal((await service.resolve()).isCanonical, false);
  assert.equal(storage.values.size, 0);
});

test('unavailable or failing storage keeps the legacy writer', async () => {
  const canonicalFetch = async () => new Response(canonicalDocument);
  const missing = new BackendActivationService({
    config, storage: null, subtle: webcrypto.subtle, fetchImpl: canonicalFetch,
  });
  assert.equal((await missing.resolve()).isCanonical, false);

  for (const storage of [
    { getItem: () => { throw new Error('read denied'); }, setItem: () => {} },
    { getItem: () => null, setItem: () => { throw new Error('write denied'); } },
    { getItem: () => null, setItem: () => {} },
  ]) {
    const service = new BackendActivationService({
      config, storage, subtle: webcrypto.subtle, fetchImpl: canonicalFetch,
    });
    assert.equal((await service.resolve()).isCanonical, false);
  }
});
